### DISCUSSION TITLE
Safetynumbers between mobile and signal-cli don't match

### DISCUSSION THREAD

**Discussion by bublath** [AsamK/signal-cli#1263] [Q&A]
I'm trying to find out which is the correct safetyNumber to use/display.
When I display the safety number for the signal-cli contact on my mobile it starts with: 2614 68799 ....
When I retrieve the safety number for my mobile via signal-cli it starts with: 326146....
The "Scannable safety number" starts with the bytes 8 1 18 34 ....
When I try to decode the QR on my mobile with https://zxing.org/w/decode it clearly looks binary and the hex bytes start with 44 a0 80 11 ....

@AsamK From previous discussions, I would have assumed that the safety numbers displayed by the mobile and in signal-cli should match - is there something wrong with the number or did I misunderstand?

**Comment by AsamK** [comment]
Yes, the safety numbers for a given contact must be the same, but only if you're registered with the same (linked) account on mobile and signal-cli.

**Reply by bublath** [reply]
So on the mobile if I use the "share" button, it generates a message with the safetynumber I see on the signal-cli side. Using this the signal-cli user could use that as an argument to trustVerified().
However I don't know how to do this the other way round. Signal-cli would somehow have to generate the number that is displayed on the mobile (or even create a QR code from it). However I don't know how to do this. Any idea? If it is encoded into the scannableSafetyNumber, I have no clue how to make a string of digits out of it.

**Reply by AsamK** [reply]
You can create a QR code from the scannable safety number, that can be scanned by the other contact's mobile app.
`signal-cli -a +XXX --output json listIdentities -n +YYY | jq .[0].scannableSafetyNumber -j | base64 -d | qrencode -t UTF8`

**Reply by bublath** [reply]
Thanks. Getting clearer. Do you also know which algorithm is used to compute the human readable representation (string of digits) from the byte array?

**Reply by AsamK** [reply]
Both the human readable numbers and scannable numbers are derived from the two fingerprints.

It's implemented here in libsignal-client: https://github.com/signalapp/libsignal/blob/main/rust/protocol/src/fingerprint.rs

**Comment by bublath** [comment]
I finally found the "error" - I was wondering why the number on my mobile was not matching - well: For some reason my screen cut of the first and the last digit per row - I was already wondering why the grouping of digits was irregular (4-5-5-4 per row).
All works as expected once I flip my screen sideways :)
